File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03734/p03734_num5_parsed.py
# importing the necessary libraries
import numpy as np

# function to solve the problem
def knapsack(N, W, weights, values):
    # creating a 2D array for memoization
    dp = np.zeros((N + 1, W + 1), dtype=int)
    
    # loop to fill the dp array
    for i in range(1, N + 1):
        for j in range(1, W + 1):
            # checking if the weight of the current item is less than or equal to the current weight
            if weights[i] <= j:
                # taking the maximum value between selecting the current item or not selecting it
                dp[i][j] = max(dp[i-1][j], dp[i-1][j-weights[i]] + values[i])
            else:
                # if the weight of the current item is greater than the current weight, not selecting it
                dp[i][j] = dp[i-1][j]
    
    # returning the maximum total value of the selected items
    return dp[N][W]

# getting input
N, W = map(int, input().split())
weights = [0]
values = [0]

# looping over the items
for i in range(N):
    w, v = map(int, input().split())
    weights.append(w)
    values.append(v)

# calling the knapsack function to get the maximum total value
max_value = knapsack(N, W, weights, values)

# printing the maximum total value
print(max_value)

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03734/p03734_num5_parsed.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Item {
    int weight;
    int value;
};

int main() {
    int N, W;
    cin >> N >> W;

    // Read the items
    vector<Item> items(N);
    for (int i = 0; i < N; i++) {
        cin >> items[i].weight >> items[i].value;
    }

    // Sort the items in increasing order of weight
    sort(items.begin(), items.end(), [](const Item& a, const Item& b) {
        return a.weight < b.weight;
    });

    // Initialize the maximum value array
    vector<int> dp(W + 1, 0);

    // Loop over the items and weights
    for (int i = 0; i < N; i++) {
        for (int j = W; j >= items[i].weight; j--) {
            // Update the maximum value
            dp[j] = max(dp[j], dp[j - items[i].weight] + items[i].value);
        }
    }

    // Print the maximum possible total value
    cout << dp[W] << endl;

    return 0;
}
